File: examples/java-calculator-junit5/src/main/java/io/cucumber/examples/junit5/calculator/RpnCalculator.java
package io.cucumber.examples.junit5.calculator;

import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

import static java.util.Arrays.asList;

public class RpnCalculator {

    private static final List<String> OPS = asList("-", "+", "*", "/");
    private final Deque<Number> stack = new LinkedList<>();

    public void push(Object arg) {
        if (OPS.contains(arg)) {
            Number y = stack.removeLast();
            Number x = stack.isEmpty() ? 0 : stack.removeLast();
            Double val = null;
            if (arg.equals("-")) {
                val = x.doubleValue() - y.doubleValue();
            } else if (arg.equals("+")) {
                val = x.doubleValue() + y.doubleValue();
            } else if (arg.equals("*")) {
                val = x.doubleValue() * y.doubleValue();
            } else if (arg.equals("/")) {
                val = x.doubleValue() / y.doubleValue();
            }
            push(val);
        } else {
            stack.add((Number) arg);
        }
    }

    public void PI() {
        push(Math.PI);
    }

    public Number value() {
        return stack.getLast();
    }

}

File: examples/calculator-kotlin-junit5/src/main/kotlin/io/cucumber/examples/calaculator/RpnCalculator.kt
package io.cucumber.examples.calaculator

class RpnCalculator {
    private val stack: ArrayDeque<Number> = ArrayDeque()

    private val OPS = setOf("+", "-", "*", "/")

    fun push(arg: Any) {
        if (arg in OPS) {
            val y = stack.removeLast()
            val x = if (stack.isEmpty()) 0 else stack.removeLast()
            val valResult = when (arg) {
                "-" -> x.toDouble() - y.toDouble()
                "+" -> x.toDouble() + y.toDouble()
                "*" -> x.toDouble() * y.toDouble()
                "/" -> x.toDouble() / y.toDouble()
                else -> throw IllegalArgumentException("Unknown operation $arg")
            }
            push(valResult)
        } else {
            stack.addLast(arg as Number)
        }
    }

    fun peek(): Number? = stack.lastOrNull()

    fun clear() = stack.clear()
}

